core/src/storage/traits/push.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{
+    ArrayLenLessEquals32,
+    KeyPtr,
+    StorageSize,
+};
+use crate::env;
+use core::marker::PhantomData;
+use ink_primitives::Key;
+
+/// Types that implement this trait can push their fields to the associated
+/// contract storage in a distributive manner.
+///
+/// # Note
+///
+/// This tries to distribute all separate fields of an entity into distinct
+/// storage cells.
+pub trait PushForward {
+    /// Pushes `self` distributed to the associated contract storage.
+    fn push_forward(&self, ptr: &mut KeyPtr);
+}
+
+/// Types that implement this trait can push their fields to an associated
+/// contract storage cell.
+///
+/// # Note
+///
+/// This tries to compactly store the entire entity's fields into a single
+/// contract storage cell.
+pub trait PushAt {
+    /// Pushes `self` packed to the contract storage cell determined by `at`.
+    fn push_at(&self, at: Key);
+}
+
+macro_rules! impl_push_for_primitive {
+    ( $($ty:ty),* $(,)? ) => {
+        $(
+            impl PushForward for $ty {
+                fn push_forward(&self, ptr: &mut KeyPtr) {
+                    <$ty as PushAt>::push_at(self, ptr.next_for::<$ty>())
+                }
+            }
+
+            impl PushAt for $ty {
+                fn push_at(&self, at: Key) {
+                    env::set_contract_storage::<$ty>(at, self)
+                }
+            }
+        )*
+    };
+}
+impl_push_for_primitive!(u8, u16, u32, u64, u128, i8, i16, i32, i64, i128);
+
+macro_rules! impl_push_for_array {
+    ( $($len:literal),* $(,)? ) => {
+        $(
+            impl<T> PushForward for [T; $len]
+            where
+                Self: ArrayLenLessEquals32,
+                T: PushForward,
+            {
+                fn push_forward(&self, ptr: &mut KeyPtr) {
+                    // TODO: Insert check to assert that the array length is
```

Comment:
ToDo for this PR or follow-up?

Reply:
The PR is not yet done so all TODOs will be worked on before merge.